File: Packages/Flip54Core/Sources/Flip54Core/PrescriptionFunction.swift
/// Returns the exercise prescription for a given card, equipment, and difficulty.
/// This is the single source of truth for all exercise/rep/hold calculations.
///
/// Suit → body region:
///   ♥ Hearts   → lower body
///   ♠ Spades   → upper body
///   ♣ Clubs    → total body
///   ♦ Diamonds → core
///   🃏 Joker   → conditioning (jumping jacks)
///
/// Within a suit, the specific exercise is picked from a pool that grows with
/// the user's available equipment, indexed deterministically by rank so a
/// workout features a varied mix while staying repeatable.
public func prescription(
    for card: Card,
    equipment: Equipment,
    difficulty: Difficulty
) -> Prescription {
    let multiplier = difficulty.multiplier

    switch card {
    case .joker:
        return .reps(exercise: .jumpingJacks, count: roundUp(40 * multiplier))

    case .standard(let suit, .ace):
        let seconds = roundToFive(60 * multiplier)
        return .hold(exercise: aceExercise(for: suit, equipment: equipment), seconds: seconds)

    case .standard(let suit, let rank):
        let exercise = movementExercise(for: suit, rank: rank, equipment: equipment)
        let basePip = rank.pipValue ?? 10  // face cards = 10
        let count = roundUp(Double(basePip) * multiplier)
        return .reps(exercise: exercise, count: count)
    }
}

// MARK: - Suit → exercise

func aceExercise(for suit: Suit, equipment: Equipment) -> Exercise {
    switch suit {
    case .hearts:   return .wallSit          // lower body
    case .spades:   return equipment.hasPullUpBar ? .deadHang : .pushUpHold  // upper body
    case .clubs:    return .plank            // total body
    case .diamonds: return .hollowBodyHold   // core
    }
}

func movementExercise(for suit: Suit, rank: Rank, equipment: Equipment) -> Exercise {
    let pool = movementPool(for: suit, equipment: equipment)
    // Deterministic per-rank index so different cards in the same suit pick
    // different exercises from the pool.
    let idx = rankIndex(rank) % pool.count
    return pool[idx]
}

/// Body-weight is always present. Add weighted moves when the user has weights;
/// add pull-ups when the user has a pull-up bar (upper body only).
func movementPool(for suit: Suit, equipment: Equipment) -> [Exercise] {
    switch suit {
    case .hearts:  // lower body
        var pool: [Exercise] = [.bodyweightSquat, .lunge, .jumpingSquat]
        if equipment.hasWeights { pool.append(.gobletSquat) }
        return pool

    case .spades:  // upper body
        var pool: [Exercise] = [.pushUp, .hinduPushUp]
        if equipment.hasWeights {
            pool += [.bicepCurl, .shoulderPress, .tricepExtension]
        }
        if equipment.hasPullUpBar { pool.append(.pullUp) }
        return pool

    case .clubs:   // total body
        var pool: [Exercise] = [.burpee, .mountainClimber, .jumpingSquat]
        if equipment.hasWeights { pool.append(.thruster) }
        return pool

    case .diamonds:  // core
        var pool: [Exercise] = [.sitUp, .russianTwist]
        if equipment.hasWeights { pool.append(.weightedSitUp) }
        return pool
    }
}

/// Ordinal of a rank in the standard 2..A sequence, used for pool indexing.
func rankIndex(_ rank: Rank) -> Int {
    switch rank {
    case .two: return 0
    case .three: return 1
    case .four: return 2
    case .five: return 3
    case .six: return 4
    case .seven: return 5
    case .eight: return 6
    case .nine: return 7
    case .ten: return 8
    case .jack: return 9
    case .queen: return 10
    case .king: return 11
    case .ace: return 12
    }
}

func roundUp(_ value: Double) -> Int {
    Int(value.rounded(.up))
}

func roundToFive(_ value: Double) -> Int {
    Int((value / 5).rounded()) * 5
}

File: android/core/src/main/kotlin/com/flip54/core/PrescriptionFunction.kt
package com.flip54.core

/**
 * Returns the exercise prescription for a given card, equipment, and difficulty.
 *
 * Suit → body region:
 *   HEARTS   → lower body
 *   SPADES   → upper body
 *   CLUBS    → total body
 *   DIAMONDS → core
 *   Joker    → conditioning (jumping jacks)
 */
fun prescription(card: Card, equipment: Equipment, difficulty: Difficulty): Prescription {
    val multiplier = difficulty.multiplier
    return when (card) {
        is Card.Joker -> Prescription.Reps(Exercise.JUMPING_JACKS, roundUp(40 * multiplier))

        is Card.Standard -> when (card.rank) {
            Rank.ACE -> {
                val seconds = roundToFive(60 * multiplier)
                Prescription.Hold(aceExercise(card.suit, equipment), seconds)
            }
            else -> {
                val exercise = movementExercise(card.suit, card.rank, equipment)
                val basePip = card.rank.pipValue ?: 10
                val count = roundUp(basePip * multiplier)
                Prescription.Reps(exercise, count)
            }
        }
    }
}

internal fun aceExercise(suit: Suit, equipment: Equipment): Exercise = when (suit) {
    Suit.HEARTS   -> Exercise.WALL_SIT
    Suit.SPADES   -> if (equipment.hasPullUpBar) Exercise.DEAD_HANG else Exercise.PUSH_UP_HOLD
    Suit.CLUBS    -> Exercise.PLANK
    Suit.DIAMONDS -> Exercise.HOLLOW_BODY_HOLD
}

internal fun movementExercise(suit: Suit, rank: Rank, equipment: Equipment): Exercise {
    val pool = movementPool(suit, equipment)
    val idx = rankIndex(rank) % pool.size
    return pool[idx]
}

internal fun movementPool(suit: Suit, equipment: Equipment): List<Exercise> = when (suit) {
    Suit.HEARTS -> buildList {
        add(Exercise.BODYWEIGHT_SQUAT)
        add(Exercise.LUNGE)
        add(Exercise.JUMPING_SQUAT)
        if (equipment.hasWeights) add(Exercise.GOBLET_SQUAT)
    }
    Suit.SPADES -> buildList {
        add(Exercise.PUSH_UP)
        add(Exercise.HINDU_PUSH_UP)
        if (equipment.hasWeights) {
            add(Exercise.BICEP_CURL)
            add(Exercise.SHOULDER_PRESS)
            add(Exercise.TRICEP_EXTENSION)
        }
        if (equipment.hasPullUpBar) add(Exercise.PULL_UP)
    }
    Suit.CLUBS -> buildList {
        add(Exercise.BURPEE)
        add(Exercise.MOUNTAIN_CLIMBER)
        add(Exercise.JUMPING_SQUAT)
        if (equipment.hasWeights) add(Exercise.THRUSTER)
    }
    Suit.DIAMONDS -> buildList {
        add(Exercise.SIT_UP)
        add(Exercise.RUSSIAN_TWIST)
        if (equipment.hasWeights) add(Exercise.WEIGHTED_SIT_UP)
    }
}

internal fun rankIndex(rank: Rank): Int = when (rank) {
    Rank.TWO   -> 0
    Rank.THREE -> 1
    Rank.FOUR  -> 2
    Rank.FIVE  -> 3
    Rank.SIX   -> 4
    Rank.SEVEN -> 5
    Rank.EIGHT -> 6
    Rank.NINE  -> 7
    Rank.TEN   -> 8
    Rank.JACK  -> 9
    Rank.QUEEN -> 10
    Rank.KING  -> 11
    Rank.ACE   -> 12
}

internal fun roundUp(value: Double): Int = kotlin.math.ceil(value).toInt()

internal fun roundToFive(value: Double): Int = (kotlin.math.round(value / 5) * 5).toInt()
